leoism/step/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -75,27 +77,44 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
       return;
     }
 
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL("/");
+      response.sendRedirect(loginUrl);
+      return;
+    }
+
+    String userEmail = userService.getCurrentUser().getEmail();
+
     double score = determineSentimentScore(comment);
 
     // Prevent users from submitting negative comments.
     if (score > 0.1) {
-      storeComment(comment, name, score);
+      storeComment(comment, name, score, userEmail);
       response.sendRedirect("/");
     } else {
       response.setContentType("text/html;");
       response.getWriter().println("<p>Please refrain from entering negative comments.</p>");
     }
   }
 
-  private void storeComment(String comment, String name, double sentimentScore) {
+  private void storeComment(String comment, String name, double sentimentScore,
+                            String email) {
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("comment", comment);
     commentEntity.setProperty("name", name);
     commentEntity.setProperty("timestamp", new Date().getTime());
     commentEntity.setProperty("sentimentScore", sentimentScore);
-
+    commentEntity.setProperty("email", email);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+    // To retrieve a proper key, the Datastore must first perform its 'put'
+    // batch actions to generate a key. Once the key is generated store it to
+    // allow identifying messages.
+    commentEntity.setProperty("key",
```

Comment:
Using an Entity builder is not possible because that belongs to the Google Cloud Datastore API and I am using the AppEngine Datastore API. Importing the Google Cloud Datastore API would cause confusion since there are classes with the same names between the two APIs.

Reply:
That's fine then. However reading the docs it seems like there's already a key assigned to the commentEntity then. Why are you setting another property called "key" on it?

Reply:
That is true, an Entity already has a property and method to retrieve a key. At first, I was not sure that iterating an iterable was possible in JSTL. Documentation said it could only iterate iterators and not iterables. Luckily, PreparedQuery has a method to convert it into an iterator.

I was able to implement it and now key is no longer necessary to store :)